content/topics/Analyze/machine-learning/supervised/Optimization_ML.md

```diff
@@ -0,0 +1,133 @@
+- It offers high accuracy
+
+#### Disadvantages: 
+- As expected, this algorithm tends to be slow on large datasets. 
+- It requires more computation and memory, as it considers all training samples at once. 
+
+### Stochastic Gradient Descent(SGD)
+Named "stochastic" due to its approach, the algorithm randomly selects one training sample in each iteration to calculate the gradient. This means the algorithm employs either a single sample entry or a small set of training samples to compute the gradient, optimizing the weights. However, since each iteration adjusts the weights using a randomly chosen sample, the cost fluctuates, and it may not consistently decrease.
+
+#### Advantages: 
+- The convergence speed of this algorithm surpasses that of BGD, making it particularly suitable for nonconvex problems like the L1 loss function.
+- In the case of very large datasets, SGD is preferred over BGD, as BGD processes the entire dataset in every iteration. However, for implementing SGD, it is recommended to use a set of training data rather than a single instance.
+
+#### Disadvantages:
+- There is a possibility that the algorithm doesn't always reach the minimum but fluctuates around it. It is important to note that with a sufficient number of iterations, the overall trend of the cost tends to decrease.
+
+### Mini Batch Gradient Descen(MGD)
+This algorithm presents a hybrid approach, blending aspects of both BGD and SGD. Unlike SGD, which involves selecting a single sample, and BGD, which considers the entire dataset, this variant of Gradient Descent operates on a fixed-length batch of training samples. By combining elements of the aforementioned algorithms, it leverages the advantages of both. In each iteration, a batch of samples is randomly chosen, and after computing the gradient on these samples, the average gradient is employed to update the weights of the cost function. It's worth noting that fluctuations may occur due to the use of a relatively small group of samples in each iteration.
+
+#### Advantages:
+- One key benefit of MBD compared to SGD is the potential performance enhancement achieved through hardware optimization of matrix operations.
+- This algorithm is favored for optimization in deep learning due to its ability to balance speed and stability effectively.
+
+#### Disadvantages: 
+- Convergence of an error cannot be assured with certainty.
+- Manually regulate the learning rate when employing MGD. If the learning rate is set too low, the convergence rate may decrease. Conversely, if the learning rate is too high, it may lead to failure in getting an absolute global or local minimum. Thus, careful control of the learning rate is essential in MGD.
+
+The image below shows the distinctions among three variants of Gradient Descent.
+<div align="center">
+  <img src="../images/GD_variants.png" width="400">
+  <p><a href="https://sweta-nit.medium.com/batch-mini-batch-and-stochastic-gradient-descent-e9bc4cacd461">Source</a></p>
```

Comment:
line 105 must be removed. 

I replaced the image from Medium with a new image I created in Python. so there's no need for this source anymore.